lib/spack/spack/package_prefs.py

```diff
@@ -182,32 +166,27 @@ def spec_externals(spec):
     from spack.util.module_cmd import get_path_from_module # NOQA: ignore=F401
 
     allpkgs = get_packages_config()
-    name = spec.name
+    names = [spec.name]
+    names += [vspec.name for vspec in spec.package.virtuals_provided()]
 
     external_specs = []
-    pkg_paths = allpkgs.get(name, {}).get('paths', None)
-    pkg_modules = allpkgs.get(name, {}).get('modules', None)
-    if (not pkg_paths) and (not pkg_modules):
-        return []
-
-    for external_spec, path in iteritems(pkg_paths):
-        if not path:
-            # skip entries without paths (avoid creating extra Specs)
+    for name in names:
+        pkg_paths = allpkgs.get(name, {}).get('paths', {})
+        pkg_modules = allpkgs.get(name, {}).get('modules', {})
+        if (not pkg_paths) and (not pkg_modules):
             continue
 
-        external_spec = spack.spec.Spec(external_spec,
-                                        external_path=canonicalize_path(path))
-        if external_spec.satisfies(spec):
-            external_specs.append(external_spec)
-
-    for external_spec, module in iteritems(pkg_modules):
-        if not module:
-            continue
+        for external_spec, path in iteritems(pkg_paths):
```

Comment:
Minor. Can we use:
```suggestion
        for external_spec, path in pkg_paths.items():
```
here and below and remove the use of `six.iteritems`? I don't expect memory request here to be the limiting factor, so Python 2 vs. Python 3 behavior shouldn't make a difference.